applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_truss_element_3D2N.h

```diff
@@ -0,0 +1,64 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:     BSD License
+//           license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Martin Fusseder, https://github.com/MFusseder
+//
+
+
+#if !defined(ADJOINT_FINITE_DIFFERENCE_TRUSS_ELEMENT_H_INCLUDED )
+#define  ADJOINT_FINITE_DIFFERENCE_TRUSS_ELEMENT_H_INCLUDED
+
+#include "adjoint_finite_difference_truss_element_linear_3D2N.h"
+
+namespace Kratos
+{
+
+class AdjointFiniteDifferenceTrussElement : public AdjointFiniteDifferenceTrussElementLinear
+{
+public:
+    KRATOS_CLASS_POINTER_DEFINITION(AdjointFiniteDifferenceTrussElement);
+
+    AdjointFiniteDifferenceTrussElement(): AdjointFiniteDifferenceTrussElementLinear()
+    {
+    }
+
+    AdjointFiniteDifferenceTrussElement(Element::Pointer pPrimalElement);
+
+    ~AdjointFiniteDifferenceTrussElement() override;
+
+    void Calculate(const Variable<Vector >& rVariable,
+                        Vector& rOutput,
+                        const ProcessInfo& rCurrentProcessInfo) override;                    
+
+    void CalculateStressDisplacementDerivative(const Variable<Vector>& rStressVariable,
+                                    Matrix& rOutput, const ProcessInfo& rCurrentProcessInfo) override;
+
+private:
+    double GetPerturbationSizeModificationFactor(const Variable<array_1d<double,3>>& rDesignVariable) override;
+
+    double CalculateReferenceLength();
+
+    double CalculateCurrentLength();
+
+    void CalculateCurrentLengthDisplacementDerivative(Vector& rDerivativeVector);
+
+    void GetDerivativePreFactor(double& rDerivativePreFactor, const ProcessInfo& rCurrentProcessInfo);
+
+    double CalculateDerivativePreFactorFX(const ProcessInfo& rCurrentProcessInfo);
```

Comment:
How are the functions `CalculateDerivativePreFactorFX` and `CalculateDerivativePreFactorPK2` different
Can you refactor to use only one function and internally decide what to do, same as what you do in other functions?

Reply:
In principle I already do this. When this pre-factor is needed first ```GetDerivativePreFactor``` is called and there internally is decided what to do. Based on this decision I call one of these two functions. Of course I can move their  code into the ```GetDerivativePreFactor```. But I am not sure if it would be nicer to have instead one lengthly function.